### PR TITLE
feat: Phase 52 — AST Safety Hook & Rule Modernization

### PR DESCRIPTION
**Pull request by chrisdevchroma**
## Summary

Replace the 8-rule AST safety avoidance table with a PreToolUse hook that auto-approves allowlisted commands at the compound-command level. Removes AST injection machinery, centralizes anti-heredoc instructions, and normalizes un-natural workflow patterns.

## Changes

- **52-01**: PreToolUse hook (`bash-safety-hook.cjs`) — compound command decomposition via character-by-character state machine, 4-layer settings merge, deny-first logic, built-in safe builtins set. Adapted from [liberzon/claude-hooks](https://github.com/liberzon/claude-hooks). 674 lines, 54 unit tests.
- **52-02**: Normalize 22 un-natural bash patterns across 6 workflow files — `printf|jq` → `echo|jq` (9 sites), `readFileSync('/dev/stdin')` → `jq` (8 sites), `process.stdin.on` → `process.argv` (5 sites)
- **52-03**: Remove AST injection machinery from `install.js` (`fillAstSafetyBlock`, markers, constants), wire hook into `settings.json` PreToolUse array (Claude-only, idempotent, appended at end), centralize anti-heredoc in `agent-shared-context.md` (both runtimes), delete `ast-safety-rules.md` template, clean `CLAUDE.md`, rename 7 tests `AST-SAFETY-*` → `BASH-HOOK-*`
- **52-04**: Remove duplicate anti-heredoc instruction from 12 agent files (now inherited via shared context), normalize `process.stdin.on` in `gsd-verifier.md`, close plan-checker dimension todo as resolved-by-hook

## Context

Claude Code's built-in bash safety heuristics trigger interactive prompts on common shell patterns (`$()`, `&&` chains, heredocs, quotes in comments) that cannot be suppressed via `permissions.allow`. See [anthropics/claude-code#30435](https://github.com/anthropics/claude-code/issues/30435).

The previous approach was an avoidance table in CLAUDE.md instructing the model to avoid these patterns. This phase replaces that with a PreToolUse hook that auto-approves at the tool-use level — transparent to the model, no constraints needed.

## Test Plan

- [x] 54/54 bash-hook.test.cjs pass (decomposition, matching, builtins, deny-first, settings layers)
- [x] 31/31 install-js.test.cjs pass (renamed BASH-HOOK-01 through -07, hook wiring, idempotency)
- [x] 0 remaining `readFileSync('/dev/stdin')`, `printf|jq`, or `process.stdin.on` in workflows
- [x] 0 per-agent anti-heredoc instructions (centralized in agent-shared-context.md)
- [x] 0 references to `GSD_AST_SAFETY_MARKER` or `fillAstSafetyBlock` in source
- [ ] Interactive verification: install, run compound command, verify no AST prompt

---
*Generated by [GSD-NG](https://github.com/chrisdevchroma/gsd-ng) from phase 52 execution*

### CODE REVIEW

File: hooks/bash-safety-hook.cjs [hunk 1]
```diff
@@ -0,0 +1,674 @@
...
+//   shift  — shift positional parameters
+//   read   — read input (from stdin, not filesystem)
+//   wait   — wait for background jobs
+//   trap   — set signal handler
+//   source — source a script (.  as alias)
+//   .      — POSIX alias for source
+const SAFE_BUILTINS = new Set([
+  'cd', 'pwd', 'echo', 'printf', 'test', '[', 'true', 'false',
+  'export', 'set', 'unset', 'local', 'return', 'exit', 'shift',
+  'read', 'wait', 'trap', 'source', '.',
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed in f58cb26. Removed `trap`, `source`, and `.` from SAFE_BUILTINS — they can execute arbitrary commands/files and must match the allowlist explicitly. Added exclusion tests (64 total now).

File: hooks/bash-safety-hook.cjs [hunk 2]
```diff
@@ -0,0 +1,674 @@
...
+  // ── Step 2: Allow check — ALL sub-commands must match ───────────────────
+  const matched = [];
+  const unmatched = [];
+
+  for (const sub of subCommands) {
+    const firstWord = sub.split(/\s+/)[0];
+
+    // Check safe builtins first
+    if (SAFE_BUILTINS.has(firstWord)) {
+      matched.push({ sub, via: 'builtin' });
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed in f58cb26. Added `extractSubshells()` that recursively extracts `$()` and backtick contents, feeding inner commands back through the allow/deny check. `echo $(rm -rf /)` now correctly decomposes to both `echo $(rm -rf /)` and `rm -rf /` — the inner `rm` fails the allowlist check so the whole command falls through to prompting. 5 new tests cover this (nested subshells, backticks, deny-through-subshell).

File: hooks/bash-safety-hook.cjs [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+
+// ── parseBashPattern ──────────────────────────────────────────────────────────
+/**
+ * Parse a Bash() permission pattern from settings.json.
+ *
+ * Pattern formats:
+ *   "Bash(git:*)"       -> prefix="git", matches commands starting with "git "
+ *   "Bash(npm run:*)"   -> prefix="npm run", matches commands starting with "npm run "
+ *   "Bash(git status)"  -> exact match (no glob)
+ *   "Bash(*)"           -> matches everything
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs:695]: Fixed in f58cb26. Docstring updated: "exact match (no glob)" → "prefix match (matches 'git status' and 'git status --porcelain')". The prefix behavior is intentional — it matches Claude Code's own permission semantics where `Bash(git status)` should match `git status` with any trailing arguments.

File: gsd-ng/workflows/plan-phase.md [hunk 4]
```diff
@@ -463,14 +463,14 @@ ls "${PHASE_DIR}"/*-PLAN.md 2>/dev/null
...
-CONTEXT_PATH=$(printf '%s\n' "$INIT" | jq -r '.context_path // empty')
-GAP_RESEARCH_PATH=$(printf '%s\n' "$INIT" | jq -r '.gap_research_path // empty')
+STATE_PATH=$(echo "$INIT" | jq -r '.state_path // empty')
+ROADMAP_PATH=$(echo "$INIT" | jq -r '.roadmap_path // empty')
+REQUIREMENTS_PATH=$(echo "$INIT" | jq -r '.requirements_path // empty')
+RESEARCH_PATH=$(echo "$INIT" | jq -r '.research_path // empty')
+VERIFICATION_PATH=$(echo "$INIT" | jq -r '.verification_path // empty')
+UAT_PATH=$(echo "$INIT" | jq -r '.uat_path // empty')
+CONTEXT_PATH=$(echo "$INIT" | jq -r '.context_path // empty')
+GAP_RESEARCH_PATH=$(echo "$INIT" | jq -r '.gap_research_path // empty')
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply gsd-ng/workflows/plan-phase.md:473]: Acknowledged but intentionally kept as `echo`. The piped data is always gsd-tools JSON output which starts with `{` — never `-n`, `-e`, or backslash sequences. The `printf` pattern was the original AST avoidance workaround; `echo` is the model-native pattern that agents naturally produce. Since the hook now handles AST prompts transparently, we normalized to what Claude would naturally generate. If a non-JSON variable is ever piped here in the future, `printf` would be warranted.

File: gsd-ng/workflows/plan-milestone-gaps.md [hunk 5]
```diff
@@ -68,7 +68,7 @@ Find highest existing phase:
 ```bash
 # Get sorted phase list, extract last one
 PHASES=$(node "$HOME/.claude/gsd-ng/bin/gsd-tools.cjs" phases list)
-HIGHEST=$(printf '%s\n' "$PHASES" | jq -r '.directories[-1]')
+HIGHEST=$(echo "$PHASES" | jq -r '.directories[-1]')
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply gsd-ng/workflows/plan-milestone-gaps.md:71]: Same as above — `$PHASES` is gsd-tools JSON output (always starts with `{`). The `echo` normalization is intentional for model-native consistency. See reply on comment #4 for full rationale.

File: hooks/bash-safety-hook.cjs [hunk 6]
```diff
@@ -0,0 +1,745 @@
...
+      let j = i + 1;
+      while (j < command.length) {
+        if (command[j] === '(') depth++;
+        else if (command[j] === ')') {
+          depth--;
+          if (depth === 0) {
+            const content = command.slice(start, j);
+            subshells.push(content);
+            // Recurse for nested subshells
+            subshells.push(...extractSubshells(content));
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed. extractSubshells() now uses a state machine tracking singleQuote/doubleQuote/backslash state when scanning for the closing ) of $() subshells. A ) inside quotes no longer prematurely closes the subshell. Test: echo $(echo ")"; rm -rf /) correctly extracts the full subshell content so the inner rm -rf / fails the allowlist check.

File: hooks/bash-safety-hook.cjs [hunk 7]
```diff
@@ -0,0 +1,745 @@
...
+
+  // Extract backtick subshells (no nesting)
+  const parts = command.split('`');
+  for (let idx = 1; idx < parts.length; idx += 2) {
+    const content = parts[idx];
+    if (content.trim()) {
+      subshells.push(content);
+      subshells.push(...extractSubshells(content));
+    }
+  }
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed. Replaced split('`') with a state-machine scanner that tracks singleQuote, doubleQuote, and backslash escape state. Escaped backticks (\`) are skipped, backticks inside single quotes are treated as literal characters, and backticks inside double quotes are correctly extracted as subshells. Tests cover all three cases.

File: hooks/bash-safety-hook.cjs [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+ * @returns {string} Normalized command (may be empty string)
+ */
+function normalizeCommand(cmd) {
+  let s = cmd.trim();
+  if (!s) return '';
+
+  // Strip heredoc body: << WORD ... WORD (simplified — remove << marker and delimiter)
+  // Match << followed by optional dash and word (quoted or unquoted)
+  s = s.replace(/<<-?\s*['"]?\w+['"]?\s*.*$/s, '').trim();
+
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed. Added stripHeredocs() pre-pass ported from liberzon/claude-hooks strip_heredocs(). Called in decomposeCommand() before splitOnOperators() so heredoc body lines are stripped and never treated as sub-commands. Tests confirm heredoc body content with deny-pattern text does not cause false denials.

File: hooks/bash-safety-hook.cjs [hunk 9]
```diff
@@ -0,0 +1,903 @@
...
+        heredocDelim = null;
+      }
+      i++;
+      continue;
+    }
+
+    // Check for heredoc marker: <<[-]?['"]?WORD['"]?
+    const m = lines[i].match(/<<-?\s*['"]?(\w+)['"]?/);
+    if (m) {
+      heredocDelim = m[1];
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed in 6052041. Widened `\w+` to `[^\s'"]+` in both `stripHeredocs` and `normalizeCommand`. Tests for `EOF-1` and `END.TXT` delimiters confirm non-word-char delimiters are now correctly parsed. This is an intentional improvement over the upstream Python which uses the same narrow `\w+`.

File: hooks/bash-safety-hook.cjs [hunk 10]
```diff
@@ -0,0 +1,903 @@
...
+
+  const homeDir = os.homedir();
+
+  // Layer 1: global settings ($CLAUDE_SETTINGS_PATH or ~/.claude/settings.json)
+  const layer1Path = env.CLAUDE_SETTINGS_PATH || path.join(homeDir, '.claude', 'settings.json');
+  const layer1 = loadSettings(layer1Path);
+
+  // Layer 2: global local settings (~/.claude/settings.local.json)
+  const layer2Path = path.join(homeDir, '.claude', 'settings.local.json');
+  const layer2 = loadSettings(layer2Path);
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed in 6052041. `loadMergedSettings` now uses `env.HOME || os.homedir()` so layer 2 resolution respects the override. The "all 4 layers" test passes `HOME` in envOverride and asserts layer 2 content is present. Integration tests use a sandboxed temp HOME via `resolveTmpDir()`.

File: tests/bash-hook.test.cjs [hunk 11]
```diff
@@ -0,0 +1,615 @@
...
+describe('BASH-HOOK-GRACEFUL: hook process graceful degradation', () => {
+  function runHook(stdin, env = {}) {
+    const mergedEnv = Object.assign({}, process.env, env);
+    const result = spawnSync(process.execPath, [HOOK_PATH], {
+      input: stdin,
+      encoding: 'utf8',
+      timeout: 5000,
+      env: mergedEnv,
+    });
+    return result;
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply tests/bash-hook.test.cjs]: Fixed in 6052041. `runHook()` now sets `HOME` to a sandbox-aware temp dir created via `resolveTmpDir()` (from helpers.cjs). Uses `before`/`after` hooks for lifecycle. Developer's `~/.claude/settings*.json` can no longer leak into integration test results.

File: tests/install-js.test.cjs [hunk 12]
```diff
@@ -986,19 +995,26 @@ test('AST-SAFETY-04: copilot local install does not create or modify CLAUDE.md',
...
+    const settingsPath = path.join(tmpDir, '.github', 'settings.json');
+    if (fs.existsSync(settingsPath)) {
+      const settings = JSON.parse(fs.readFileSync(settingsPath, 'utf8'));
+      const preToolUse = (settings.hooks && settings.hooks.PreToolUse) || [];
+      const hasBashSafety = preToolUse.some(entry =>
+        entry.hooks && entry.hooks.some(h => h.command && h.command.includes('bash-safety-hook.cjs'))
+      );
+      assert.ok(!hasBashSafety,
+        'copilot settings.json must NOT contain bash-safety-hook.cjs in PreToolUse (BASH-HOOK-04)');
+    }
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply tests/install-js.test.cjs]: Fixed in 6052041. Comment updated to "does NOT wire bash-safety-hook into settings.json" and added explicit file-existence guardrail assertion: `assert.ok(!fs.existsSync(hookFilePath))` for `.github/hooks/bash-safety-hook.cjs`. This serves as a guardrail for future runtimes, not just Copilot.

File: hooks/bash-safety-hook.cjs [hunk 13]
```diff
@@ -0,0 +1,903 @@
...
+ *   Layer 1: $CLAUDE_SETTINGS_PATH || ~/.claude/settings.json
+ *   Layer 2: ~/.claude/settings.local.json
+ *   Layer 3: $CLAUDE_PROJECT_DIR/.claude/settings.json
+ *   Layer 4: $CLAUDE_PROJECT_DIR/.claude/settings.local.json
+ */
+
+'use strict';
+
+// ── Kill switch — MUST be FIRST, before any stdin read or file access ─────────
+if (process.env.GSD_DISABLE_BASH_HOOK === '1') {
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed in 6052041. Kill switch moved inside the `require.main === module` block. `require()`ing the module no longer calls `process.exit(0)` — only CLI execution honors the kill switch. Existing subprocess test still passes since it runs the module as main.

File: hooks/bash-safety-hook.cjs [hunk 14]
```diff
@@ -0,0 +1,903 @@
...
+      i++;
+      continue;
+    }
+
+    // Check for heredoc marker: <<[-]?['"]?WORD['"]?
+    const m = lines[i].match(/<<-?\s*['"]?([^\s'"]+)['"]?/);
+    if (m) {
+      heredocDelim = m[1];
+    }
+
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed in 6052041. Added `isInsideQuotes(line, idx)` helper that does a left-to-right scan tracking single/double quote state. `stripHeredocs` now checks if the `<<` position is inside a quoted region before accepting a heredoc delimiter. Three new tests confirm: `echo "text <<EOF"\nrm -rf /\nEOF` does NOT strip the `rm -rf /` line (the `<<` is inside quotes).

File: hooks/bash-safety-hook.cjs [hunk 15]
```diff
@@ -0,0 +1,986 @@
...
+      current += ch;
+      i++;
+      continue;
+    }
+
+    // ) — exit subshell
+    if (ch === ')' && depth > 0) {
+      depth--;
+      current += ch;
+      i++;
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs:342]: Fixed in 6052041. Added bare `(` tracking when `depth > 0`: `if (ch === '(' && depth > 0) { depth++; ... }`. This matches the Python upstream's behavior. `$( (cmd1; cmd2) )` and `$(( 1 + 2 ))` now balance correctly. Two new tests confirm no premature depth close.

File: tests/install-js.test.cjs [hunk 16]
```diff
@@ -961,23 +966,27 @@ test('AST-SAFETY-03: claude local install is idempotent — re-running does not
...
+    assert.strictEqual(bashSafetyEntries.length, 1,
+      'bash-safety-hook.cjs must appear exactly once in PreToolUse after two installs (BASH-HOOK-03). ' +
+      'Found: ' + bashSafetyEntries.length + ' entries');
   } finally {
     fs.rmSync(tmpDir, { recursive: true, force: true });
   }
 });
 
-// ── AST-SAFETY-04: copilot install does NOT touch CLAUDE.md ──────────────────
+// ── BASH-HOOK-04: copilot install does NOT create bash-safety-hook ─────────────
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply tests/install-js.test.cjs]: Fixed in 6052041. Comment updated to "does NOT wire bash-safety-hook into settings.json" and added explicit `!fs.existsSync()` assertion for the hook file in the Copilot hooks directory.

File: .planning/quick/260407-w82-fix-parsing-edge-cases-in-bash-safety-ho/SUMMARY.md [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+  added: []
+  patterns:
+    - "State machine with singleQuote/doubleQuote/backslash tracking for all bash parsers"
+    - "stripHeredocs() pre-pass pattern: strip heredoc bodies before operator splitting"
+
+key-files:
+  created: []
+  modified:
+    - gsd-ng/hooks/bash-safety-hook.cjs
+    - gsd-ng/tests/bash-hook.test.cjs
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply .planning/quick/260407-w82-fix-parsing-edge-cases-in-bash-safety-ho/SUMMARY.md]: Fixed in 6052041. The entire `.planning/` directory was removed from the submodule via `git rm -r` — planning docs belong in the workspace, not the submodule. Added `.planning/` to `.gitignore` to prevent recurrence.

File: hooks/bash-safety-hook.cjs [hunk 18]
```diff
@@ -0,0 +1,1024 @@
...
+      continue;
+    }
+
+    // Check for heredoc marker: <<[-]?['"]?WORD['"]?
+    const m = lines[i].match(/<<-?\s*['"]?([^\s'"]+)['"]?/);
+    if (m) {
+      const heredocPos = lines[i].indexOf('<<');
+      if (!isInsideQuotes(lines[i], heredocPos)) {
+        heredocDelim = m[1];
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed in 1298d09. The suggested `(?!<)` lookahead alone is **insufficient** — the regex engine can match the 2nd+3rd `<` of `<<<` at position offset +1, bypassing the lookahead. Verified by test:

```
// Copilot's (?!<) only:
cat <<<"$x"  →  match at index 5, captures $x  ← BYPASS STILL WORKS
cat <<<word  →  match at index 5, captures word ← BYPASS STILL WORKS
```

Fix uses **both** `(?<!<)` lookbehind + `(?!<)` lookahead:

```js
const HEREDOC_START_RE = /(?<!<)<<(?!<)-?\s*(?:'([^'\n]+)'|"([^"\n]+)"|([^\s'"]+))/;
```

- `(?<!<)` prevents matching at position 5 (preceded by `<` at position 4)
- `(?!<)` prevents matching at position 4 (followed by `<` at position 6)
- Together: only standalone `<<` is treated as a heredoc

This is an improvement over the upstream Python (`\w+`) which is immune to `<<<` by accident but can't match non-word delimiters like `EOF-1` or `END.TXT`.

Extracted as shared `HEREDOC_START_RE` constant used by both `stripHeredocs()` and `normalizeCommand()`. 8 new regression tests in `BASH-HOOK-HERESTRING` suite (123 total now).

File: hooks/bash-safety-hook.cjs [hunk 19]
```diff
@@ -0,0 +1,1024 @@
...
+      continue;
+    }
+
+    // ( — track nested parens inside subshell
+    if (ch === '(' && depth > 0) {
+      depth++;
+      current += ch;
+      i++;
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Confirmed — this is a porting bug from Python → JS. Fixed in d4751fa.

**Root cause:** Python consumes `$(` as a unit with `i += 2`. The JS port used `i++`, leaving `(` for the next iteration where it hits the `if (ch === '(' && depth > 0)` handler and double-increments depth. After `$(cmd)`, depth is 1 instead of 0 — all subsequent `&&`/`;`/`|` operators are treated as inside a subshell and never split.

**Exploit:** `echo $(git status) && rm -rf /` stays as one segment. `echo` is in `SAFE_BUILTINS` → the whole compound command is auto-approved, including `rm -rf /`.

**Fix:** Both `$(` handlers (outside quotes at line 330 and inside double quotes at line 303) now use `i += 2` and `current += ch + next`, matching the Python upstream.

6 new regression tests in `BASH-HOOK-SUBSHELL-DEPTH` suite including a `decide()` security assertion. 129 total tests pass.

File: hooks/bash-safety-hook.cjs [hunk 20]
```diff
@@ -0,0 +1,1024 @@
...
+
+    // $( — enter subshell
+    if (ch === '$' && next === '(') {
+      depth++;
+      current += ch;
+      i++;
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Already fixed in e125672 — see reply on the identical finding at line 330. Both `$(` handlers (outside quotes and inside double quotes) now use `i += 2` and `current += ch + next`, matching the Python upstream. 6 regression tests in `BASH-HOOK-SUBSHELL-DEPTH` suite.

File: tests/bash-hook.test.cjs [hunk 21]
```diff
@@ -0,0 +1,887 @@
...
+    const result = splitOnOperators('$( (echo a; echo b) )');
+    assert.equal(result.length, 1,
+      'bare parens inside $() must not cause premature depth-0 exit');
+  });
+
+  test('handles arithmetic $(( )) without double-increment issues', () => {
+    const result = splitOnOperators('echo $(( 1 + 2 ))');
+    assert.equal(result.length, 1,
+      'arithmetic $(( )) must remain single segment');
+  });
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply tests/bash-hook.test.cjs:100]: Already added in e125672 — `BASH-HOOK-SUBSHELL-DEPTH` suite has 6 regression tests including:
- `$() then && splits into two segments`
- `$() then ; splits into two segments`
- `nested $() then && splits correctly`
- `multiple $() then && splits correctly`
- `$() in assignment context then && splits correctly`
- `decide() does NOT auto-approve chained dangerous command`

All verified as true regression tests (fail with old code, pass with fix).

File: tests/bash-hook.test.cjs [hunk 22]
```diff
@@ -0,0 +1,887 @@
...
+describe('BASH-HOOK-HEREDOC: heredoc body lines are not treated as sub-commands', () => {
+  const { stripHeredocs } = hook;
+
+  test('stripHeredocs removes heredoc body lines', () => {
+    const input = 'cat <<EOF\nhello world\nrm -rf /\nEOF';
+    const result = stripHeredocs(input);
+    assert.ok(!result.includes('rm -rf /'), 'heredoc body must be stripped');
+    assert.ok(result.includes('cat <<EOF'), 'heredoc marker line must be preserved');
+  });
+
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply tests/bash-hook.test.cjs:422]: Already added in e125672 — `BASH-HOOK-HERESTRING` suite has 8 regression tests covering all `<<<` variants:
- `cat <<<"$x"\nrm -rf /`
- `cat <<<word\nrm -rf /`
- `cat <<<'word'\nrm -rf /`
- `cat <<< word\nrm -rf /`
- `<<<<\nrm -rf /`
- `<<<` then real heredoc on next line
- `<<<EOF-1` (non-word value)
- `<<<END.TXT` (dotted value)

All verified as true regression tests.

File: hooks/bash-safety-hook.cjs [hunk 23]
```diff
@@ -0,0 +1,1042 @@
...
+    const rest = s.slice(endVal).trimStart();
+    if (!rest) break;  // standalone assignment — stop stripping, keep as-is
+    s = rest;
+  }
+
+  // Strip trailing redirections: >, >>, <, 2>&1, &>, /dev/null, etc.
+  // Simple approach: strip common redirection patterns from the end
+  // Remove: 2>&1, &>, &>>, >>, >, <, followed by optional filename
+  s = s.replace(/\s+[0-9]*>>?\s*\S+$/, '');
+  s = s.replace(/\s+[0-9]*>&[0-9]+$/, '');
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs:449]: Valid finding. Fixed in f05d829 by **removing SAFE_BUILTINS entirely**.

SAFE_BUILTINS was a gsd-ng addition not present in the Python upstream (liberzon/claude-hooks). It auto-approved builtins (`echo`, `cd`, `read`, `printf`, etc.) without any allowlist entry. Combined with redirection stripping in `normalizeCommand()`, this allowed commands like `echo "x" > ~/.bashrc` and `read var < /etc/shadow` to be auto-approved — a filesystem bypass.

**Resolution:** All commands (including builtins) now go through the same allowlist path. GSD-NG's `settings-sandbox.json` template already includes `Bash(echo *)`, `Bash(cd *)`, etc. — so sandbox users see no change. Non-sandbox users must explicitly allowlist builtins they want auto-approved, matching the Python upstream's behavior.

5 regression tests in `BASH-HOOK-NO-BUILTINS` suite:
- `echo hello` → passthrough without allowlist
- `cd /tmp` → passthrough without allowlist
- `echo "x" > ~/.bashrc` → passthrough (redirection bypass blocked)
- `read var < /etc/shadow` → passthrough (redirection bypass blocked)
- `echo hello` → approve WITH explicit `Bash(echo:*)` allowlist

120 bash-hook + 31 install tests pass.

File: hooks/bash-safety-hook.cjs [hunk 24]
```diff
@@ -0,0 +1,1012 @@
+#!/usr/bin/env node
+/**
+ * bash-safety-hook.cjs — PreToolUse hook for compound bash command allowlist matching
+ *
+ * Decomposes compound bash commands (&&, ||, ;, |, $() subshells) into individual
+ * sub-commands and auto-approves when ALL sub-commands match the user's allowlist
+ * or are known-safe shell builtins.
```
[the remark this answers is not in the document]
Comment (chrisdevchroma) [reply hooks/bash-safety-hook.cjs]: Fixed in 5ac1cd0. Updated the file header doc comment to reflect the SAFE_BUILTINS removal.
